Add `async_without_await` rule

jkolarik-paylocity opened the pull request:
Adds `async_without_await` [#5082](https://github.com/realm/SwiftLint/issues/5082)

JiriKaminek commented on CHANGELOG.md:
```diff
@@ -10,7 +10,9 @@
 #### Enhancements
 
-* None.
+* Adds `noAsyncWithoutAwait` opt-in rule that check if a async function call at least one async function.  
```
Do we want to fix warnings in this PR? There is so many of them it would be nice to do that in single PR.

SimplyDanny replied:
```suggestion
* Add new `no_async_without_await` opt-in rule that check if an `async` function/property/subscript/initializer declaration contains at least one `await`.  
```

jkolarik-paylocity commented on Source/SwiftLintFramework/Configuration+CommandLine.swift:
```diff
@@ -241,6 +241,7 @@ extension Configuration {
             linters.asyncMap(visit)
     }
 
+    // swiftlint:disable:next no_async_without_await
```
It seems that this method doesn't need to be `async`. Suppress or remove `async`?

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoAsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,120 @@
+
+    static let description = RuleDescription(
+        identifier: "no_async_without_await",
+        name: "No sync without await",
```
```suggestion
        name: "No Async Without Await",
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoAsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,120 @@
+    static let description = RuleDescription(
+        identifier: "no_async_without_await",
+        name: "No sync without await",
+        description: "Function should not be async if doesn't use await",
```
```suggestion
        description: "Declaration should not be async if it doesn't use await",
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoAsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,120 @@
+        identifier: "no_async_without_await",
+        name: "No sync without await",
+        description: "Function should not be async if doesn't use await",
+        kind: .idiomatic,
```
`.lint` is probably a better fitting category.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoAsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,120 @@
+}
+private extension NoAsyncWithoutAwaitRule {
+    private struct FuncInfo {
+        var awaitCount = 0
```
Can be a `Bool`.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoAsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,120 @@
+            }
+        }
+
+        // swiftlint:disable:next unused_parameter
+        override func visitPost(_ node: AwaitExprSyntax) {
```
```suggestion
        override func visitPost(_: AwaitExprSyntax) {
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Idiomatic/NoAsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,120 @@
+
+private extension FunctionDeclSyntax {
+    var asyncSpecifier: TokenSyntax? {
+        guard let asyncSpecifier = signature.effectSpecifiers?.asyncSpecifier else {
+            return nil
+        }
+
+        return asyncSpecifier
```
```suggestion
        signature.effectSpecifiers?.asyncSpecifier
```

SimplyDanny reviewed (changes requested):
Thank you for this first implementation!

Apart from my inline comments I have a few remarks:

* The current implementation only covers functions. Think about adding support for other declarations supporting `async`.
* Automatic fixes could be offered by this rule.
* I presume the rule will currently fail to flag the `async` on `f` in the following example:
  ```swift
  func f() async {
    func g() async {}
    let x = { await g() }
  }
  ```

jkolarik-paylocity commented:
Hi @SimplyDanny,
- I have added support for:
  - `init() async`
  - `get async`
  - `let x: () async -> Void = {}`
 - Added `SwiftSyntaxCorrectableRule`.
 - Renamed `no_async_without_await` to `async_without_await`.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRule.swift:
```diff
@@ -1,6 +1,8 @@
 import SwiftLintCore
 import SwiftSyntax
 
+// swiftlint:disable file_length
```
We usually move the examples into a separate file if the actual rule becomes too long.

SimplyDanny commented on CHANGELOG.md:
```diff
@@ -10,7 +10,9 @@
 #### Enhancements
 
-* None.
+* Add new `async_without_await` opt-in rule that check if an `async` function/property/subscript/initializer declaration contains at least one `await`.  
```
Reading that again, we could skip the details and be a bit more concise:
```suggestion
* Add new `async_without_await` opt-in rule that checks if an `async` declaration contains at least one `await`.  
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,418 @@
+    }
+
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        private var awaitCount = Stack<FuncInfo>()
```
```suggestion
        private var functionScopes = Stack<FuncInfo>()
```

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,418 @@
+                return
+            }
+
+            if !info.containsAwait {
```
Condition can be moved into the `guard`.

SimplyDanny reviewed (changes requested):
Well done!

Again a few nits and some remarks:

* Please check the findings in OSS repositories. [This violation](https://github.com/brave/brave-ios/blob/398f8b763aa88cdc23289138863d62f05b2c2a23/Tests/ClientTests/User%20Scripts/MockScriptsViewController.swift#L63:52) is a false positive for example.
* The current implementation doesn't respect `async let x = ...` constructs.

msrutek-paylocity commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,418 @@
+        }
+
+        private func checkViolation() {
+            guard let info = awaitCount.pop(), let asyncToken = info.asyncToken  else {
```
```suggestion
            guard let info = awaitCount.pop(), let asyncToken = info.asyncToken else {
```
nit: double space

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRuleExamples.swift:
```diff
@@ -0,0 +1,328 @@
+        }
+        """),
+        Example("""
+        func test() ↓async {
```
This example should actually not trigger the rule. `async let x = ...` is only allowed in an `async` context.

jkolarik-paylocity replied:
You are right. I will fix it.

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRuleExamples.swift:
```diff
@@ -0,0 +1,338 @@
+            while let foo = bar() {}
+        }
+        """),
+        Example("func foo(bar: () ↓async -> Void = {}) { }"),
```
This looks strange. For a parameter, only because the default value isn't `async` isn't a sufficient indicator that the whole closure doesn't need to be `async`.

jkolarik-paylocity replied:
So what do you suggest? Ignore it if it is a default argument.

msrutek-paylocity reviewed (approved):
👏

DominikBucher12 reviewed (approved):
🧡

SimplyDanny commented on Source/SwiftLintBuiltInRules/Rules/Lint/AsyncWithoutAwaitRule.swift:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        private func checkViolation() {
+            guard
+                let info = functionScopes.pop(),
+                let asyncToken = info.asyncToken,
+                !info.containsAwait
+            else {
+                return
+            }
```
```suggestion
            guard let info = functionScopes.pop(),
                  let asyncToken = info.asyncToken,
                  !info.containsAwait else {
                return
            }
```

SimplyDanny commented:
Thanks again for contributing this rule and the fast responses, @jkolarik-paylocity!

Very impressive to see how you structured the rule and how you handled all the special cases in efficient ways. Well done! 👍
